Tests/ASCollectionViewThrashTests.mm

```diff
@@ -0,0 +1,184 @@
+//  Changes after 4/13/2017 are: Copyright (c) Pinterest, Inc.  All rights reserved.
+//  Licensed under Apache 2.0: http://www.apache.org/licenses/LICENSE-2.0
+//
+
+#import <XCTest/XCTest.h>
+#import <AsyncDisplayKit/AsyncDisplayKit.h>
+#import <AsyncDisplayKit/ASCollectionView.h>
+#import <stdatomic.h>
+
+#import "ASThrashUtility.h"
+
+@interface ASCollectionViewThrashTests : XCTestCase
+
+@end
+
+@implementation ASCollectionViewThrashTests  {
+  // The current update, which will be logged in case of a failure.
+  ASThrashUpdate *_update;
+  BOOL _failed;
+}
+
+- (void)tearDown {
+  if (_failed && _update != nil) {
+    NSLog(@"Failed update %@: %@", _update, _update.logFriendlyBase64Representation);
+  }
+  _failed = NO;
+  _update = nil;
+}
+
+// NOTE: Despite the documentation, this is not always called if an exception is caught.
+- (void)recordFailureWithDescription:(NSString *)description inFile:(NSString *)filePath atLine:(NSUInteger)lineNumber expected:(BOOL)expected {
+  _failed = YES;
+  [super recordFailureWithDescription:description inFile:filePath atLine:lineNumber expected:expected];
+}
+
+- (void)verifyDataSource:(ASThrashDataSource *)ds {
+  CollectionView *collectionView = ds.collectionView;
+  NSArray <ASThrashTestSection *> *data = [ds data];
+  for (NSInteger i = 0; i < collectionView.numberOfSections; i++) {
+    XCTAssertEqual([collectionView numberOfItemsInSection:i], data[i].items.count);
+
+    for (NSInteger j = 0; j < [collectionView numberOfItemsInSection:i]; j++) {
+      NSIndexPath *indexPath = [NSIndexPath indexPathForItem:j inSection:i];
+      ASThrashTestItem *item = data[i].items[j];
+      ASThrashTestNode *node = (ASThrashTestNode *)[collectionView nodeForItemAtIndexPath:indexPath];
+      XCTAssertEqualObjects(node.item, item, @"Wrong node at index path %@", indexPath);
+    }
+  }
+}
+
+#pragma mark Test Methods
+
+// Disabled temporarily due to issue where cell nodes are not marked invisible before deallocation.
+- (void)testInitialDataRead {
+  ASThrashDataSource *ds = [[ASThrashDataSource alloc] initCollectionViewDataSourceWithData:[ASThrashTestSection sectionsWithCount:kInitialSectionCount]];
+  [self verifyDataSource:ds];
+}
+
+/// Replays the Base64 representation of an ASThrashUpdate from "ASThrashTestRecordedCase" file
+- (void)testRecordedThrashCase {
+  NSURL *caseURL = [[NSBundle bundleForClass:[self class]] URLForResource:@"ASThrashTestRecordedCase" withExtension:nil subdirectory:@"TestResources"];
+  NSString *base64 = [NSString stringWithContentsOfURL:caseURL encoding:NSUTF8StringEncoding error:NULL];
+
+  _update = [ASThrashUpdate thrashUpdateWithBase64String:base64];
+  if (_update == nil) {
+    return;
+  }
+
+  ASThrashDataSource *ds = [[ASThrashDataSource alloc] initCollectionViewDataSourceWithData:_update.oldData];
+  // why is ds.collectionView.test_enableSuperUpdateCallLogging available on table view but now collection view>?
```

Comment:
We just didn't need it before. May worth adding it if you think it'd be helpful.